File: tests/test/test-complete-workflow.js
// test-complete-workflow.js - Final comprehensive test of the entire workflow system

const fetch = require('node-fetch');
const readline = require('readline');

const BASE_URL = 'http://localhost:8002';

// Test credentials
const testUser = {
    email: 'test@example.com',
    password: 'password123'
};

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question) {
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            resolve(answer);
        });
    });
}

async function testCompleteWorkflow() {
    console.log('🔬 Testing COMPLETE workflow with manual confirmation...\n');
    
    try {
        // Step 1: Login
        console.log('📝 Step 1: Logging in...');
        const loginResponse = await fetch(`${BASE_URL}/api/auth/login`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(testUser)
        });
        
        if (!loginResponse.ok) {
            throw new Error(`Login failed: ${loginResponse.statusText}`);
        }
        
        const cookies = loginResponse.headers.get('set-cookie');
        const sessionCookie = cookies ? cookies.split(',').find(c => c.includes('session=')) : null;
        
        if (!sessionCookie) {
            throw new Error('No session cookie received');
        }
        
        const cookieHeader = sessionCookie.split(';')[0];
        console.log('✅ Login successful!\n');
        
        // Step 2: Generate workflow preview
        console.log('🤖 Step 2: Generating workflow preview...');
        const mcpResponse = await fetch(`${BASE_URL}/api/mcp/chat`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': cookieHeader
            },
            body: JSON.stringify({
                message: "Create a marketing email for our new fitness app and send it to slakshanand1105@gmail.com",
                ai_service: "in_house"
            })
        });
        
        if (!mcpResponse.ok) {
            throw new Error(`MCP request failed: ${mcpResponse.statusText}`);
        }
        
        const mcpData = await mcpResponse.json();
        console.log('✅ Workflow preview generated!');
        console.log(`   Type: ${mcpData.type}`);
        console.log(`   Status: ${mcpData.status}`);
        console.log(`   Has workflow preview: ${!!mcpData.workflow_preview}`);
        
        if (mcpData.workflow_preview) {
            console.log(`   Title: ${mcpData.workflow_preview.title}`);
            console.log(`   Description: ${mcpData.workflow_preview.description}`);
            console.log(`   Steps: ${mcpData.workflow_preview.steps?.length || 0}`);
            console.log(`   Credits: ${mcpData.workflow_preview.estimated_credits}`);
            console.log(`   Recipient: ${mcpData.workflow_preview.recipient}`);
            
            if (mcpData.workflow_preview.workflow_json) {
                console.log('\\n📋 WORKFLOW DETAILS:');
                console.log(`   Type: ${mcpData.workflow_preview.workflow_json.type}`);
                console.log(`   Recipient: ${mcpData.workflow_preview.workflow_json.recipient}`);
                console.log(`   Content Type: ${mcpData.workflow_preview.workflow_json.content_type}`);
                
                // Step 3: Manual confirmation test
                console.log('\\n🎯 Step 3: Testing Manual Confirmation...');
                console.log('In the frontend, you should see a dialog with:');
                console.log('- Workflow title and description');
                console.log('- Step details');
                console.log('- Credit cost');
                console.log('- CANCEL and CONFIRM & EXECUTE buttons');
                console.log('');
                
                const userDecision = await ask('Did you see the workflow preview dialog? (y/n): ');
                
                if (userDecision.toLowerCase() === 'y') {
                    console.log('✅ Dialog display working correctly!');
                    
                    const confirmChoice = await ask('Do you want to test workflow execution? (y/n): ');
                    
                    if (confirmChoice.toLowerCase() === 'y') {
                        console.log('\\n🚀 Step 4: Testing workflow execution...');
                        const confirmResponse = await fetch(`${BASE_URL}/api/workflow/confirm`, {
                            method: 'POST',
                            headers: {
                                'Content-Type': 'application/json',
                                'Cookie': cookieHeader
                            },
                            body: JSON.stringify({
                                workflow_json: mcpData.workflow_preview.workflow_json
                            })
                        });
                        
                        if (!confirmResponse.ok) {
                            const errorText = await confirmResponse.text();
                            console.log(`❌ Workflow execution failed: ${confirmResponse.status} - ${errorText}`);
                        } else {
                            const confirmData = await confirmResponse.json();
                            console.log('✅ Workflow execution completed!');
                            console.log(`   Success: ${confirmData.success}`);
                            console.log(`   Message: ${confirmData.message}`);
                            
                            if (confirmData.execution_details) {
                                console.log(`   Details: ${JSON.stringify(confirmData.execution_details, null, 2)}`);
                            }
                        }
                    } else {
                        console.log('⏭️  Skipping execution test');
                    }
                } else {
                    console.log('❌ Dialog not displaying - there may be a frontend issue');
                }
            } else {
                console.log('⚠️  No workflow_json found');
            }
        } else {
            console.log('❌ No workflow preview generated');
        }
        
    } catch (error) {
        console.error('❌ Test failed:', error.message);
    } finally {
        rl.close();
    }
}

async function main() {
    console.log('🧪 COMPLETE WORKFLOW TEST SUITE\\n');
    console.log('This test will check:');
    console.log('✓ Backend workflow generation');
    console.log('✓ Frontend dialog display');
    console.log('✓ Manual user confirmation');
    console.log('✓ Real automation execution');
    console.log('='.repeat(60));
    console.log('');
    
    console.log('📋 INSTRUCTIONS:');
    console.log('1. Make sure both servers are running:');
    console.log('   - Backend: http://localhost:8002');
    console.log('   - Frontend: http://localhost:3001');
    console.log('2. Open the frontend at: http://localhost:3001/dashboard/agents/sam/chat');
    console.log('3. Type this message: "Create a marketing email for our new fitness app and send it to slakshanand1105@gmail.com"');
    console.log('4. Watch for the workflow preview dialog');
    console.log('5. Answer the prompts in this terminal');
    console.log('='.repeat(60));
    console.log('');
    
    const ready = await ask('Are you ready to start the test? (y/n): ');
    
    if (ready.toLowerCase() === 'y') {
        await testCompleteWorkflow();
    } else {
        console.log('Test cancelled');
        rl.close();
    }
}

main().catch(console.error);

File: tests/test/test_complete_workflow.py
#!/usr/bin/env python3
"""
Test Complete Frontend-Backend Flow
Simulate the exact flow the user will experience
"""

import requests
import json
import time

BASE_URL = "http://localhost:8002"
FRONTEND_URL = "http://localhost:3000"

def test_complete_workflow():
    """Test the complete AI + Email workflow through frontend"""
    print("🧪 Testing Complete Frontend Workflow")
    print("=" * 50)
    
    # Step 1: Login via frontend
    print("🔐 Step 1: Logging in...")
    login_response = requests.post(
        f"{FRONTEND_URL}/api/auth/login",
        json={"email": "aitest@example.com", "password": "testpass123"}
    )
    
    if login_response.status_code != 200:
        print(f"❌ Login failed: {login_response.status_code}")
        return False
    
    cookies = login_response.cookies
    print("✅ Login successful")
    
    # Step 2: Send initial message that should trigger AI service selection
    print("\n🤖 Step 2: Sending AI + Email request...")
    
    message_payload = {
        "message": "Use AI to generate a sale pitch for our healthy ice cream and send email to slakshanand1105@gmail.com",
        "agentId": "test-agent",
        "agentConfig": {
            "name": "Sam",
            "role": "Personal Assistant"
        }
    }
    
    chat_response = requests.post(
        f"{FRONTEND_URL}/api/chat/mcpai",
        json=message_payload,
        cookies=cookies
    )
    
    if chat_response.status_code != 200:
        print(f"❌ Chat request failed: {chat_response.status_code}")
        print(f"Error: {chat_response.text}")
        return False
    
    chat_result = chat_response.json()
    print("✅ Initial message sent")
    print(f"Response status: {chat_result.get('status', 'unknown')}")
    print(f"Action required: {chat_result.get('action_required', 'none')}")
    
    # Check if AI service selection was triggered
    if chat_result.get('action_required') == 'select_ai_service':
        print("✅ AI service selection triggered correctly")
        ai_services = chat_result.get('ai_service_options', [])
        print(f"Available services: {[s.get('name') for s in ai_services]}")
        
        # Step 3: Select AI service (simulate clicking "Inhouse")
        print("\n⚙️ Step 3: Selecting AI service...")
        service_message = {
            "message": "service:inhouse Use AI to generate a sale pitch for our healthy ice cream and send email to slakshanand1105@gmail.com",
            "agentId": "test-agent",
            "agentConfig": {
                "name": "Sam",
                "role": "Personal Assistant"
            }
        }
        
        service_response = requests.post(
            f"{FRONTEND_URL}/api/chat/mcpai",
            json=service_message,
            cookies=cookies
        )
        
        if service_response.status_code != 200:
            print(f"❌ Service selection failed: {service_response.status_code}")
            return False
        
        service_result = service_response.json()
        print("✅ AI service selected")
        print(f"Response status: {service_result.get('status', 'unknown')}")
        print(f"Action required: {service_result.get('action_required', 'none')}")
        
        # Check if workflow confirmation is needed
        if 'workflow' in str(service_result) or service_result.get('action_required') == 'confirm_workflow':
            print("✅ Workflow generated and ready for confirmation")
            
            # Step 4: Confirm workflow execution
            print("\n✅ Step 4: Confirming workflow...")
            confirm_payload = {
                "agentId": "test-agent",
                "confirmed": True,
                "workflow_json": service_result.get('workflow_json', {})
            }
            
            confirm_response = requests.post(
                f"{FRONTEND_URL}/api/chat/mcpai/confirm",
                json=confirm_payload,
                cookies=cookies
            )
            
            if confirm_response.status_code != 200:
                print(f"❌ Workflow confirmation failed: {confirm_response.status_code}")
                print(f"Error: {confirm_response.text}")
                return False
            
            confirm_result = confirm_response.json()
            print("✅ Workflow confirmed and executed")
            print(f"Success: {confirm_result.get('success')}")
            print(f"Response: {confirm_result.get('response', 'No response')[:100]}...")
            
            return True
        else:
            print(f"⚠️ Expected workflow confirmation but got: {service_result}")
            return False
    else:
        print(f"⚠️ Expected AI service selection but got: {chat_result}")
        return False

if __name__ == "__main__":
    success = test_complete_workflow()
    
    print("\n" + "=" * 50)
    if success:
        print("🎉 COMPLETE WORKFLOW TEST PASSED!")
        print("✅ Frontend → Backend → AI Selection → Workflow → Email")
    else:
        print("❌ Workflow test failed - see details above")
